server/channels/app/platform/web_broadcast_hook.go

```diff
@@ -0,0 +1,52 @@
+package platform
+
+import "github.com/mattermost/mattermost/server/public/model"
+
+type BroadcastHook struct {
+	HasChanges func(msg *model.WebSocketEvent, webConn *WebConn, args map[string]any) bool
+	Process    func(msg *model.WebSocketEvent, webConn *WebConn, args map[string]any) *model.WebSocketEvent
```

Comment:
I broke up the hook into two methods to avoid allocating unnecessary `model.WebSocketEvent`s every time an event is published with a hook attached. This seems important for cases like the `posted` event where a post in a busy channel may cause thousands of events to be broadcast when only a few users are mentioned.

I could've made it so that `BroadcastHook.Process` is required to return a new object if it makes changes, but I was wanting to make it so that hook implementors won't have to worry about that. It's only a shallow copy, so they do still have to use `AddWithCopy` and not mutate the inner fields, but this feel slightly nicer at least.

I'm happy to change this though if anyone has any strong opinions